Panic when use api dht/provide.

<!--
Welcome to the go-ipfs bug tracker. This is for you! Please read, and then delete this text before posting it.

If you haven't yet searched the issue tracker for an existing report concerning your issue, please do so now.

The go-ipfs issues are only for bug reports and directly actionable feature requests. Read https://github.com/ipfs/community/blob/master/contributing.md#reporting-issues if your issue doesn't fit either of those categories.

If you have a *SUPPORT QUESTION*, please direct it to our forum at https://discuss.ipfs.io.

Read https://github.com/ipfs/go-ipfs/blob/master/docs/github-issue-guide.md if you are not sure how to fill in this issue.
-->

#### Version information:
<!--
Output From `ipfs version --all`

Please check dist.ipfs.io for a newer version of go-ipfs
and update if neccessary, then check again if the problem persists.
-->
go-ipfs version: 0.4.17-
Repo version: 7
System version: amd64/windows
Golang version: go1.10.3
#### Type:
Bug

#### Description:
<!--
This is where you get to tell us what went wrong or what feature you need. When doing so, please make sure to include *all* relevant information.

When requesting a feature, please be sure to include:

* Your motivation. Why do you need the feature?
* How the feature should work.

When reporting a bug, please try to include:

* What you were doing when you experienced the bug.
* Any error messages you saw, *where* you saw them, and what you believe may have caused them (if you have any ideas).
* When possible, steps to reliably produce the bug.
-->
config changes:
```
{
  "API": {
    "HTTPHeaders": {
     "Access-Control-Allow-Headers": [
        "X-Requested-With",
        "Range",
		"User-Agent"
      ],
      "Access-Control-Allow-Methods": [
        "GET"
      ],
      "Access-Control-Allow-Origin": [
        "*"
      ]
    }
  },
"Addresses": {
"Swarm": []
}
  "Swarm": {
    "DisableNatPortMap": true
  }
}
```

I use fetch() in browser to send provide request to ipfs daemon. This error happen:
```
panic: send on closed channel

goroutine 2628 [running]:
gx/ipfs/QmZ383TySJVeZWzGnWui6pRcKyYZk9VkKTuW7tmKRWk5au/go-libp2p-routing/notifications.PublishQueryEvent(0x109d060, 0xc04287d840, 0xc04228ba80)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmZ383TySJVeZWzGnWui6pRcKyYZk9VkKTuW7tmKRWk5au/go-libp2p-routing/notifications/query.go:48 +0x151
gx/ipfs/QmTktQYCKzQjhxF6dk5xJPRuhHn3JBiKGvMLoiDy1mYmxC/go-libp2p-kad-dht.(*dhtQueryRunner).addPeerToQuery(0xc0425b52c0, 0xc04283a2d0, 0x22)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmTktQYCKzQjhxF6dk5xJPRuhHn3JBiKGvMLoiDy1mYmxC/go-libp2p-kad-dht/query.go:186 +0xee
gx/ipfs/QmTktQYCKzQjhxF6dk5xJPRuhHn3JBiKGvMLoiDy1mYmxC/go-libp2p-kad-dht.(*dhtQueryRunner).queryPeer(0xc0425b52c0, 0x10a4e60, 0xc043011740, 0xc0422e4360, 0x22)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmTktQYCKzQjhxF6dk5xJPRuhHn3JBiKGvMLoiDy1mYmxC/go-libp2p-kad-dht/query.go:307 +0x3bc
gx/ipfs/QmTktQYCKzQjhxF6dk5xJPRuhHn3JBiKGvMLoiDy1mYmxC/go-libp2p-kad-dht.(*dhtQueryRunner).spawnWorkers.func1(0x10a4e60, 0xc043011740)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmTktQYCKzQjhxF6dk5xJPRuhHn3JBiKGvMLoiDy1mYmxC/go-libp2p-kad-dht/query.go:218 +0x57
gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go.func1(0xc0428acb60, 0xc043011740, 0xc0430117a0)
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:112 +0x43
created by gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go
        /home/steb/projects/go/src/github.com/ipfs/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:111 +0x216
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,7 +5,7 @@ os:
 language: go
 
 go:
-  - 1.9.x
+  - 1.11.x
 
 install:
   - make deps

--- a/notifications/query.go
+++ b/notifications/query.go
@@ -88,12 +88,12 @@ func PublishQueryEvent(ctx context.Context, ev *QueryEvent) {
 }
 
 func (qe *QueryEvent) MarshalJSON() ([]byte, error) {
-	out := make(map[string]interface{})
-	out["ID"] = peer.IDB58Encode(qe.ID)
-	out["Type"] = int(qe.Type)
-	out["Responses"] = qe.Responses
-	out["Extra"] = qe.Extra
-	return json.Marshal(out)
+	return json.Marshal(map[string]interface{}{
+		"ID":        peer.IDB58Encode(qe.ID),
+		"Type":      int(qe.Type),
+		"Responses": qe.Responses,
+		"Extra":     qe.Extra,
+	})
 }
 
 func (qe *QueryEvent) UnmarshalJSON(b []byte) error {

--- a/notifications/query.go
+++ b/notifications/query.go
@@ -3,14 +3,16 @@ package notifications
 import (
 	"context"
 	"encoding/json"
+	"sync"
 
 	peer "github.com/libp2p/go-libp2p-peer"
 	pstore "github.com/libp2p/go-libp2p-peerstore"
 )
 
-const RoutingQueryKey = "RoutingQueryEvent"
-
 type QueryEventType int
+
+// Number of events to buffer.
+var QueryEventBufferSize = 16
 
 const (
 	SendingQuery QueryEventType = iota
@@ -30,25 +32,59 @@ type QueryEvent struct {
 	Extra     string
 }
 
-func RegisterForQueryEvents(ctx context.Context, ch chan<- *QueryEvent) context.Context {
-	return context.WithValue(ctx, RoutingQueryKey, ch)
+type routingQueryKey struct{}
+type eventChannel struct {
+	mu  sync.Mutex
+	ctx context.Context
+	ch  chan<- *QueryEvent
+}
+
+// waitThenClose is spawned in a goroutine when the channel is registered. This
+// safely cleans up the channel when the context has been canceled.
+func (e *eventChannel) waitThenClose() {
+	<-e.ctx.Done()
+	e.mu.Lock()
+	close(e.ch)
+	// 1. Signals that we're done.
+	// 2. Frees memory (in case we end up hanging on to this for a while).
+	e.ch = nil
+	e.mu.Unlock()
+}
+
+// send sends an event on the event channel, aborting if either the passed or
+// the internal context expire.
+func (e *eventChannel) send(ctx context.Context, ev *QueryEvent) {
+	e.mu.Lock()
+	// Closed.
+	if e.ch == nil {
+		e.mu.Unlock()
+		return
+	}
+	// in case the passed context is unrelated, wait on both.
+	select {
+	case e.ch <- ev:
+	case <-e.ctx.Done():
+	case <-ctx.Done():
+	}
+	e.mu.Unlock()
+}
+
+func RegisterForQueryEvents(ctx context.Context) (context.Context, <-chan *QueryEvent) {
+	ch := make(chan *QueryEvent, QueryEventBufferSize)
+	ech := &eventChannel{ch: ch, ctx: ctx}
+	go ech.waitThenClose()
+	return context.WithValue(ctx, routingQueryKey{}, ech), ch
 }
 
 func PublishQueryEvent(ctx context.Context, ev *QueryEvent) {
-	ich := ctx.Value(RoutingQueryKey)
+	ich := ctx.Value(routingQueryKey{})
 	if ich == nil {
 		return
 	}
 
-	ch, ok := ich.(chan<- *QueryEvent)
-	if !ok {
-		return
-	}
-
-	select {
-	case ch <- ev:
-	case <-ctx.Done():
-	}
+	// We *want* to panic here.
+	ech := ich.(*eventChannel)
+	ech.send(ctx, ev)
 }
 
 func (qe *QueryEvent) MarshalJSON() ([]byte, error) {

--- a/notifications/query_test.go
+++ b/notifications/query_test.go
@@ -1,0 +1,44 @@
+package notifications
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+)
+
+func TestEventsCancel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx, events := RegisterForQueryEvents(ctx)
+	goch := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		for i := 0; i < 100; i++ {
+			PublishQueryEvent(ctx, &QueryEvent{Extra: fmt.Sprint(i)})
+		}
+		close(goch)
+		for i := 100; i < 1000; i++ {
+			PublishQueryEvent(ctx, &QueryEvent{Extra: fmt.Sprint(i)})
+		}
+	}()
+	go func() {
+		defer wg.Done()
+		i := 0
+		for e := range events {
+			if i < 100 {
+				if e.Extra != fmt.Sprint(i) {
+					t.Errorf("expected %d, got %s", i, e.Extra)
+				}
+			}
+			i++
+		}
+		if i < 100 {
+			t.Errorf("expected at least 100 events, got %d", i)
+		}
+	}()
+	<-goch
+	cancel()
+	wg.Wait()
+}

